include/util/settings/shortcuts.h

```diff
@@ -0,0 +1,162 @@
+void setShortcuts(ShortcutType type, QList<QKeySequence> shortcuts);
+
+class ShortcutsItem {
+ public:
+  ShortcutsItem(QString display_name = QString(""), ShortcutsItem* parent = 0);
+  ShortcutsItem(QString display_name, ShortcutsItem* parent, util::settings::shortcuts::ShortcutType type);
+  ShortcutsItem(QString name, QString display_name, ShortcutsItem* parent, util::settings::shortcuts::ShortcutType type);
+  ~ShortcutsItem();
+
+  QAction* createQAction(QObject* parent, Qt::ShortcutContext context = Qt::WindowShortcut);
+  QAction* createQAction(QObject* parent, QIcon icon, Qt::ShortcutContext context = Qt::WindowShortcut);
+  QString name() const;
+  QString displayName() const;
+  ShortcutsItem* parent() const;
+  QList<QKeySequence> shortcuts() const;
+  QString displayShortcuts() const;
+  bool deleteShortcut(const QKeySequence& shortcut);
+  bool addShortcut(QKeySequence& shortcut);
+  void setConflict(bool conflict);
+  bool hasConflict() const;
+  util::settings::shortcuts::ShortcutType type() const;
+  void updateShortcutsForActions();
+  bool isCategory() const;
+
+  QList<ShortcutsItem*> children;
+
+ private:
+  QList<QPointer<QAction>> actions_;
+  QList<QKeySequence> shortcuts_;
+  QString display_shortcuts_;
+  QString name_;
+  QString display_name_;
+  ShortcutsItem* parent_;
+  bool conflict_;
+  util::settings::shortcuts::ShortcutType type_;
+  bool is_category_;
+};
+
+class ShortcutsModel : public QAbstractItemModel {
+ public:
+  static const int COLUMN_INDEX_NAME = 0;
+  static const int COLUMN_INDEX_SHORTCUTS = 1;
+  static const int CATEGORY_ROLE = Qt::UserRole;
+  static const int TYPE_ROLE = Qt::UserRole + 1;
+  static const int SHORTCUTS_ROLE = Qt::UserRole + 2;
+
+  ~ShortcutsModel();
+
+  virtual QModelIndex index(int row, int column, const QModelIndex& parent = QModelIndex()) const override;
+  virtual QModelIndex parent(const QModelIndex& index) const override;
+  virtual int rowCount(const QModelIndex& parent = QModelIndex()) const override;
+  virtual int columnCount(const QModelIndex& parent = QModelIndex()) const override;
+  virtual QVariant data(const QModelIndex& index, int role = Qt::DisplayRole) const override;
+  QVariant headerData(int section, Qt::Orientation orientation, int role = Qt::DisplayRole) const override;
+  static ShortcutsModel* getShortcutsModel();
+  QAction* createQAction(util::settings::shortcuts::ShortcutType type, QObject* parent, Qt::ShortcutContext context = Qt::WindowShortcut);
+  QAction* createQAction(util::settings::shortcuts::ShortcutType type, QObject* parent, QIcon icon, Qt::ShortcutContext context = Qt::WindowShortcut);
+  void addShortcut(util::settings::shortcuts::ShortcutType type, QKeySequence& shortcut);
+  void removeShortcut(util::settings::shortcuts::ShortcutType type, const QKeySequence& shortcut);
+  ShortcutsItem* itemFromIndex(const QModelIndex& index) const;
+  QModelIndex indexFromItem(ShortcutsItem* item) const;
```

Comment:
Due to how Qt interfaces are defined I would need to use const_cast somewhere if I add const here - I'm leaning towards leaving it as it is - how do you feel about it?